### PR TITLE
FIX: bots could generate errors when slug generation method is encoded

### PR DESCRIPTION
**Pull request by SamSaffron**
When slug generation method is encoded (non default) then bots could
cause errors in the logs for urls containing special chars.

ó for example in a URL can be requested in a valid ASCII-8BIT string, and
later when joined to UTF-8 would result in encoding issues.

Fix here ensures we force encoding correctly for outlier cases.

Browser tend to always encode these chars, hence we did not notice this.

### CODE REVIEW

File: app/helpers/application_helper.rb [hunk 1]
```diff
@@ -282,12 +291,13 @@ def crawlable_meta_data(opts = nil)
     opts[:twitter_summary_large_image] =
       get_absolute_image_url(opts[:twitter_summary_large_image]) if opts[:twitter_summary_large_image].present?
 
-    # Add opengraph & twitter tags
     result = []
+    result << (+"i-think-it’s-time-to-drop-the-expertise-system/")
```
Comment (SamSaffron) [inline app/helpers/application_helper.rb:295]: yikes ... oops removing

File: spec/helpers/application_helper_spec.rb [hunk 2]
```diff
@@ -161,7 +161,7 @@ def preload_link(url)
 
     context "when dark theme is present" do
       before do
-        dark_theme = Theme.create(
+        _dark_theme = Theme.create(
```
Comment (oblakeerickson) [inline spec/helpers/application_helper_spec.rb:164]: Was this intentional?

Comment (oblakeerickson) [reply spec/helpers/application_helper_spec.rb:164]: Oh. Is the _ signifying it is an unused local variable?

Comment (SamSaffron) [reply spec/helpers/application_helper_spec.rb:164]: yeah... that var is not used at all the `_` prefix denotes that

File: spec/helpers/application_helper_spec.rb [hunk 3]
```diff
@@ -418,6 +418,15 @@ def preload_link(url)
   end
 
   describe 'crawlable_meta_data' do
+
+    it 'Supports ASCII URLs with odd chars' do
+      result = helper.crawlable_meta_data(
+        url: -((+"http://localhost/ión").force_encoding("ASCII-8BIT"))
```
Comment (SamSaffron) [inline spec/helpers/application_helper_spec.rb:424]: - = freeze a string
+ = the opposite

File: app/helpers/application_helper.rb [hunk 4]
```diff
@@ -282,12 +289,12 @@ def crawlable_meta_data(opts = nil)
     opts[:twitter_summary_large_image] =
       get_absolute_image_url(opts[:twitter_summary_large_image]) if opts[:twitter_summary_large_image].present?
 
-    # Add opengraph & twitter tags
     result = []
     result << tag(:meta, property: 'og:site_name', content: SiteSetting.title)
     result << tag(:meta, property: 'og:type', content: 'website')
 
-    result = generate_twitter_card_metadata(opts, result)
+    generate_twitter_card_metadata(result, opts)
```
Comment (oblakeerickson) [inline app/helpers/application_helper.rb:296]: Don't we still need the `result` variable here?

File: app/helpers/application_helper.rb [hunk 5]
```diff
@@ -316,7 +323,7 @@ def crawlable_meta_data(opts = nil)
     result.join("\n")
   end
 
-  def generate_twitter_card_metadata(opts, result)
+  private def generate_twitter_card_metadata(result, opts)
```
[the remark this answers is not in the document]
Comment (oblakeerickson) [reply app/helpers/application_helper.rb:326]: Oh nice. Ya first time seeing private used that way :)

File: app/helpers/application_helper.rb (hunk 4 above)
Comment (SamSaffron) [reply app/helpers/application_helper.rb:296]: It was confusing previously, it was amending the input and also returning it, maybe we should rename and add a ! for clarity

File: spec/helpers/application_helper_spec.rb (hunk 3 above)
Comment (CvX) [inline spec/helpers/application_helper_spec.rb:424]: Just a tiny bit less magic? 😉 

```suggestion
        url: (+"http://localhost/ión").force_encoding("ASCII-8BIT").freeze
```

**Review by CvX — APPROVED** [review]

**Pull request merged**